sql: can't expandPlan twice!

The daily logic tests are failing. Guessing this is a recent change to subquery handling. See https://groups.google.com/a/cockroachlabs.com/forum/?hl=en#!topic/daily/Ki7HOhrM1qY.

```
panic: can't expandPlan twice!

goroutine 501 [running]:
panic(0x15c6760, 0xc820a412e0)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/cockroachdb/cockroach/sql.(*selectTopNode).expandPlan(0xc820a95240, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/select_top.go:65 +0xa1
github.com/cockroachdb/cockroach/sql.(*subqueryPlanVisitor).VisitPre(0xc820dc0ed8, 0x7f66f3c423e0, 0xc8209f4540, 0xc820e0a4e8, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/subquery.go:200 +0xcd
github.com/cockroachdb/cockroach/sql/parser.WalkExpr(0x7f66f3b30708, 0xc820dc0ed8, 0x7f66f3c423e0, 0xc8209f4540, 0x0, 0x0, 0x1862900)
	/go/src/github.com/cockroachdb/cockroach/sql/parser/walk.go:407 +0x56
github.com/cockroachdb/cockroach/sql/parser.(*ComparisonExpr).Walk(0xc820809500, 0x7f66f3b30708, 0xc820dc0ed8, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/parser/walk.go:164 +0xce
github.com/cockroachdb/cockroach/sql/parser.WalkExpr(0x7f66f3b30708, 0xc820dc0ed8, 0x7f66f3c3bd60, 0xc820809500, 0x7f66f3c3bd60, 0xc820809500, 0xc8209f4300)
	/go/src/github.com/cockroachdb/cockroach/sql/parser/walk.go:410 +0x98
github.com/cockroachdb/cockroach/sql/parser.(*OrExpr).Walk(0xc820861680, 0x7f66f3b30708, 0xc820dc0ed8, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/parser/walk.go:260 +0xc0
github.com/cockroachdb/cockroach/sql/parser.WalkExpr(0x7f66f3b30708, 0xc820dc0ed8, 0x7f66f3c3cd88, 0xc820861680, 0x7f66f3c3cd88, 0xc820861680, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/parser/walk.go:410 +0x98
github.com/cockroachdb/cockroach/sql/parser.(*AndExpr).Walk(0xc8208616b0, 0x7f66f3b30708, 0xc820dc0ed8, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/parser/walk.go:58 +0x5f
github.com/cockroachdb/cockroach/sql/parser.WalkExpr(0x7f66f3b30708, 0xc820dc0ed8, 0x7f66f3bbc4a8, 0xc8208616b0, 0x7f66f3bbc4a8, 0xc8208616b0, 0x7f66f3c80000)
	/go/src/github.com/cockroachdb/cockroach/sql/parser/walk.go:410 +0x98
github.com/cockroachdb/cockroach/sql.(*planner).expandSubqueryPlans(0xc820dc0890, 0x7f66f3bbc4a8, 0xc8208616b0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/subquery.go:223 +0x107
github.com/cockroachdb/cockroach/sql.(*scanNode).expandPlan(0xc820d52580, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/scan.go:118 +0x81
github.com/cockroachdb/cockroach/sql.(*indexJoinNode).expandPlan(0xc8208095e0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/join.go:156 +0x6b
github.com/cockroachdb/cockroach/sql.(*selectNode).expandPlan(0xc8202f77a0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/select.go:427 +0x904
github.com/cockroachdb/cockroach/sql.(*selectTopNode).expandPlan(0xc820b0dec0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/select_top.go:71 +0xf4
github.com/cockroachdb/cockroach/sql.(*planner).makePlan(0xc820dc0890, 0x7f66f3bb6848, 0xc820a45590, 0x829f01, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/plan.go:215 +0xe2
github.com/cockroachdb/cockroach/sql.(*Executor).execStmt(0xc8203042a0, 0x7f66f3bb6848, 0xc820a45590, 0xc820dc0890, 0xc820a45501, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:923 +0x93
github.com/cockroachdb/cockroach/sql.(*Executor).execStmtInOpenTxn(0xc8203042a0, 0x7f66f3bb6848, 0xc820a45590, 0xc820dc0890, 0x101, 0xc820dc0828, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:826 +0x545
github.com/cockroachdb/cockroach/sql.(*Executor).execStmtsInCurrentTxn(0xc8203042a0, 0xc820a402a0, 0x1, 0x1, 0xc820dc0890, 0xc820dc0828, 0x1f10101, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:605 +0x5da
github.com/cockroachdb/cockroach/sql.runTxnAttempt(0xc8203042a0, 0xc820dc0890, 0x1, 0xc820dc0828, 0xc820d1a7a0, 0xc820a402a0, 0x1, 0x1, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:535 +0x18d
github.com/cockroachdb/cockroach/sql.(*Executor).execRequest.func1(0xc820121b80, 0xc820d1a7a0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:444 +0x34a
github.com/cockroachdb/cockroach/client.(*Txn).Exec(0xc820121b80, 0xc8208f0101, 0x1453d9b52c1267c0, 0x0, 0xc8208f19a0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/client/txn.go:527 +0x2eb
github.com/cockroachdb/cockroach/sql.(*Executor).execRequest(0xc8203042a0, 0x7f66f3c3ad50, 0xc8210d4780, 0xc820dc0800, 0xc82075aa72, 0x1a5, 0x0, 0x0, 0x0, 0x7f66f3bbaf00)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:449 +0x962
github.com/cockroachdb/cockroach/sql.(*Executor).ExecuteStatements(0xc8203042a0, 0x7f66f3c3ad50, 0xc8210d4780, 0xc820dc0800, 0xc82075aa72, 0x1a5, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:338 +0xe5
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).executeStatements(0xc820e40480, 0x7f66f3c3ad50, 0xc8210d4780, 0xc82075aa72, 0x1a5, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:648 +0xb2
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).handleSimpleQuery(0xc820e40480, 0x7f66f3c3ad50, 0xc8210d4780, 0xc820e404a8, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:325 +0xf2
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).serve(0xc820e40480, 0xc820688880, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:281 +0x1095
github.com/cockroachdb/cockroach/sql/pgwire.(*Server).ServeConn(0xc821055b00, 0x7f66f3bbaef0, 0xc82023e900, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:224 +0x99b
github.com/cockroachdb/cockroach/server.(*Server).Start.func8.1(0x7f66f3bf33e8, 0xc82032e000)
	/go/src/github.com/cockroachdb/cockroach/server/server.go:361 +0x42
github.com/cockroachdb/cockroach/util.(*Server).ServeWith.func1(0xc8203360b0, 0x7f66f3bf33e8, 0xc82032e000, 0xc8210360c0)
	/go/src/github.com/cockroachdb/cockroach/util/net.go:131 +0x62
created by github.com/cockroachdb/cockroach/util.(*Server).ServeWith
	/go/src/github.com/cockroachdb/cockroach/util/net.go:133 +0x333
```

[Archive.zip](https://github.com/cockroachdb/cockroach/files/293486/Archive.zip)

--- a/sql/Makefile
+++ b/sql/Makefile
@@ -1,3 +1,3 @@
 .PHONY: bigtest
 bigtest:
-	go test -bigtest -timeout 24h -run '^TestLogic$$'
+	go test -bigtest -timeout 24h $(TESTFLAGS) -run '^TestLogic$$'

--- a/sql/select.go
+++ b/sql/select.go
@@ -338,16 +338,6 @@ func (p *planner) SelectClause(parsed *parser.SelectClause, orderBy parser.Order
 }
 
 func (s *selectNode) expandPlan() error {
-	// Expand the sub-query plans in the sub-expressions, if any.
-	if err := s.planner.expandSubqueryPlans(s.filter); err != nil {
-		return err
-	}
-	for _, e := range s.render {
-		if err := s.planner.expandSubqueryPlans(e); err != nil {
-			return err
-		}
-	}
-
 	// Get the ordering for index selection (if any).
 	var ordering columnOrdering
 	var grouping bool
@@ -418,20 +408,23 @@ func (s *selectNode) expandPlan() error {
 		s.source.plan = plan
 	}
 
-	// If the source node is not a scanNode, expand its plan.
-	// (If it is a scanNode, it is already "expanded" and
-	// we do not want to expand the sub-queries in its filters
-	// here because they are shared with the selectNode and the selectNode
-	// already cares of expanding the sub-queries.)
-	if _, isScan := s.source.plan.(*scanNode); !isScan {
-		if err := s.source.plan.expandPlan(); err != nil {
+	s.ordering = s.computeOrdering(s.source.plan.Ordering())
+
+	// Expand the sub-query plans in the local sub-expressions, if any.
+	// This must be done for filters after index selection and splitting
+	// the filter, since part of the filter may have landed in the source
+	// scanNode and will be expanded there.
+	if err := s.planner.expandSubqueryPlans(s.filter); err != nil {
+		return err
+	}
+	for _, e := range s.render {
+		if err := s.planner.expandSubqueryPlans(e); err != nil {
 			return err
 		}
 	}
 
-	s.ordering = s.computeOrdering(s.source.plan.Ordering())
-
-	return nil
+	// Expand the source node.
+	return s.source.plan.expandPlan()
 }
 
 // initFrom initializes the table node, given the parsed select expression

--- a/sql/testdata/subquery
+++ b/sql/testdata/subquery
@@ -305,3 +305,14 @@ query B
 INSERT INTO xyz (x, y, z) VALUES (13, 11, 12) RETURNING (y IN (SELECT y FROM xyz))
 ----
 true
+
+# check that residual filters are not expanded twice
+query ITTT
+EXPLAIN(VERBOSE) SELECT x FROM xyz WHERE x IN (SELECT x FROM xyz);
+----
+0 select                          +x,unique
+1 render/filter  (x)@xyz          +x,unique
+2 scan           xyz@primary  -   +x,unique
+3 select                          +x,unique
+4 render/filter  (x)@xyz          +x,unique
+5 scan           xyz@primary      +x,unique

